modules/persistence/src/services/metadata/associations/index.ts

```diff
@@ -0,0 +1,147 @@
+import { ToC, ToCInsertions, TocOrderInsertions, traverseAndMerge } from '../ToC';
+
+export interface AssociatedProduct {
+  name: string;
+  versions: string[];
+}
+
+export interface SharedMetadata {
+  project: string;
+  branch: string;
+  associated_products?: AssociatedProduct[];
+  toctree: ToC;
+  toctreeOrder: any[];
+  [key: string]: any;
+}
+
+// TODO: move the branch/repobranch interfaces into their own file, or into a seperate abstraction?
+interface BranchEntry {
+  name: string;
+  gitBranchName: string;
+  [key: string]: any;
+}
+
+export interface ReposBranchesDocument {
+  repoName: string;
+  project: string;
+  branches: BranchEntry[];
+  [key: string]: any;
+}
+
+// Queries pool*.repos_branches for any entries for the given project and branch from a metadata entry.
+const repoBranchesEntry = async (project, branch) => {
+  const db = await pool();
+  return db.collection('repos_branches').find({ branches: { $elemMatch: { gitBranchName: branch } }, project });
+};
+
+// Queries pool*.repos_branches for all entries for associated_products in a shared metadata entry
+const allAssociatedRepoBranchesEntries = async (metadata: SharedMetadata) => {
+  const { associated_products = [] } = metadata;
+  const associatedProductNames = associated_products.map((a) => a.name);
+  const db = await pool();
+  const entries = await db
+    .collection('repos_branches')
+    .find({ project: { $in: associatedProductNames } })
+    .toArray();
+  return entries as unknown as ReposBranchesDocument[];
+};
+
+const mapRepoBranches = (repoBranches: ReposBranchesDocument[]) =>
+  Object.fromEntries(
+    repoBranches.map((entry) => {
+      const branches = Object.fromEntries(entry.branches.map((branch) => [branch.name, branch]));
+      return [entry.project, branches];
+    })
+  );
+
+const hasAssociations = (metadata) => !!metadata.associated_products?.length;
+
+const sharedMetadataEntry = async (metadata): Promise<SharedMetadata> => {
+  try {
+    const snooty = await db();
+    const entry = await snooty
+      .collection('metadata')
+      .find({
+        $elemMatch: { associated_products: metadata.project },
+      })
+      .sort({ build_id: -1 })
+      .limit(1)
+      .toArray();
+    return entry[0] as unknown as SharedMetadata;
+  } catch (error) {
+    console.log(`Error at time of querying for umbrella metadata entry: ${error}`);
+    throw error;
+  }
+};
+
+// Convert our cursor from the shared ToC aggregation query into a series of ToC objects
+const shapeToCsCursor = async (
+  tocCursor: AggregationCursor,
+  repoBranchesMap
+): Promise<{ tocInsertions: ToCInsertions; tocOrderInsertions: TocOrderInsertions }> => {
+  let tocInsertions, tocOrderInsertions;
+  await tocCursor.forEach((doc) => {
+    // TODO: If we want staging builds with embedded versions, it needs to be added here
+    if (repoBranchesMap[doc._id.project][doc._id.branch]) {
+      tocInsertions[doc._id.project][doc._id.branch] = doc.most_recent.tocTree;
```

Comment:
seems like these likes will throw errors when reading project from tocInsertions and tocOrderInsertions (undefined values in line 84)